crates/control/src/game_state_filter.rs

```diff
@@ -229,18 +250,55 @@ impl State {
         }
     }
 
-    fn construct_filtered_game_state(
+    fn construct_hulks_filtered_game_state(
         &self,
         game_controller_state: &GameControllerState,
         cycle_start_time: SystemTime,
         ball_detected_far_from_kick_off_point: bool,
         config: &GameStateFilterParameters,
     ) -> FilteredGameState {
-        let is_in_sub_state = game_controller_state.sub_state.is_some();
         let opponent_is_kicking_team = matches!(
             game_controller_state.kicking_team,
             Team::Opponent | Team::Uncertain
         );
+        self.construct_filtered_game_state(
+            game_controller_state,
+            opponent_is_kicking_team,
+            cycle_start_time,
+            ball_detected_far_from_kick_off_point,
+            config,
+        )
+    }
+
+    fn construct_opponent_filtered_game_state(
+        &self,
+        game_controller_state: &GameControllerState,
+        cycle_start_time: SystemTime,
+        ball_detected_far_from_kick_off_point: bool,
+        config: &GameStateFilterParameters,
+    ) -> FilteredGameState {
+        let hulks_is_kicking_team = matches!(
+            game_controller_state.kicking_team,
+            Team::Hulks | Team::Uncertain
```

Comment:
Shouldn't only one of these cases include the Team::Uncertain?

Reply:
This value comes from the game controller state message where it is set to either `Hulks or `Opponent` during parsing.
AFAIK the `Uncertain` state is only used in the filtered game state when a whistle is heard during play.
Since we have no reliable way to determine which team will be the kicking team, it is set to `Uncertain` until we receive the true value from the game controller a few seconds later.

It would probably be more correct to assert this here instead of coercing `Uncertain` to the opponent team having kickoff.

Reply:
If I understand you correctly @knoellle, you mean the same thing as @MaikRe? Now replaced by only `Team::Hulks`